pkg/postage/listener/listener.go

```diff
@@ -0,0 +1,284 @@
+		priceOracleAddress:  priceOracleAddress,
+		quit:                make(chan struct{}),
+	}
+}
+
+func (l *listener) filterQuery(from, to *big.Int) ethereum.FilterQuery {
+	return ethereum.FilterQuery{
+		FromBlock: from,
+		ToBlock:   to,
+		Addresses: []common.Address{
+			l.postageStampAddress,
+			l.priceOracleAddress,
+		},
+		Topics: [][]common.Hash{
+			{
+				batchCreatedTopic,
+				batchTopupTopic,
+				batchDepthIncreaseTopic,
+				priceUpdateTopic,
+			},
+		},
+	}
+}
+
+func (l *listener) processEvent(e types.Log, updater postage.EventUpdater) error {
+	switch e.Topics[0] {
+	case batchCreatedTopic:
+		c := &batchCreatedEvent{}
+		err := transaction.ParseEvent(&postageStampABI, "BatchCreated", c, e)
+		if err != nil {
+			return err
+		}
+		return updater.Create(
+			c.BatchId[:],
+			c.Owner.Bytes(),
+			c.NormalisedBalance,
+			c.Depth,
+		)
+	case batchTopupTopic:
+		c := &batchTopUpEvent{}
+		err := transaction.ParseEvent(&postageStampABI, "BatchTopUp", c, e)
+		if err != nil {
+			return err
+		}
+		return updater.TopUp(
+			c.BatchId[:],
+			c.NormalisedBalance,
+		)
+	case batchDepthIncreaseTopic:
+		c := &batchDepthIncreaseEvent{}
+		err := transaction.ParseEvent(&postageStampABI, "BatchDepthIncrease", c, e)
+		if err != nil {
+			return err
+		}
+		return updater.UpdateDepth(
+			c.BatchId[:],
+			c.NewDepth,
+			c.NormalisedBalance,
+		)
+	case priceUpdateTopic:
+		c := &priceUpdateEvent{}
+		err := transaction.ParseEvent(&priceOracleABI, "PriceUpdate", c, e)
+		if err != nil {
+			return err
+		}
+		return updater.UpdatePrice(
+			c.Price,
+		)
+	default:
+		return errors.New("unknown event")
+	}
+}
+
+func (l *listener) Listen(from uint64, updater postage.EventUpdater) <-chan struct{} {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-l.quit
+		cancel()
+	}()
+
+	synced := make(chan struct{})
+	closeOnce := new(sync.Once)
+	paged := make(chan struct{}, 1)
+	paged <- struct{}{}
+
+	l.wg.Add(1)
+	listenf := func() error {
+		defer l.wg.Done()
+		for {
+			select {
+			case <-paged:
+				// if we paged then it means there's more things to sync on
+			case <-time.After(chainUpdateInterval):
+			case <-l.quit:
+				return nil
+			}
+			to, err := l.ev.BlockNumber(ctx)
+			if err != nil {
+				return err
+			}
+
+			if to < tailSize {
+				// in a test blockchain there might be not be enough blocks yet
+				continue
+			}
+
+			// consider to-tailSize as the "latest" block we need to sync to
+			to = to - tailSize
+
+			if to < from {
+				// if the blockNumber is actually less than what we already, it might mean the backend is not synced or some reorg scenario
+				continue
+			}
+
+			// do some paging (sub-optimal)
+			if to-from > blockPage {
+				paged <- struct{}{}
+				to = from + blockPage
+			} else {
+				closeOnce.Do(func() { close(synced) })
+			}
+
+			events, err := l.ev.FilterLogs(ctx, l.filterQuery(big.NewInt(int64(from)), big.NewInt(int64(to))))
```

Comment:
this might break with certain endpoints if there are too many (postage) events in that range.